logger/src/rate_limiting.rs

```diff
@@ -0,0 +1,89 @@
+use std::net::IpAddr;
+
+use tonic::{
+    metadata::{KeyAndValueRef, MetadataMap},
+    transport::server::TcpConnectInfo,
+    Status,
+};
+use tower::BoxError;
+use tower_governor::{key_extractor::KeyExtractor, GovernorError};
+
+/// The interval at which the rate limiter refreshes one slot in milliseconds.
+pub const REFRESH_INTERVAL: u64 = 500;
+/// The quota of requests that can be received before rate limiting is applied.
+pub const BURST_SIZE: u32 = 6;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct TonicPeerIpKeyExtractor;
+
+impl KeyExtractor for TonicPeerIpKeyExtractor {
+    type Key = IpAddr;
+
+    fn name(&self) -> &'static str {
+        "peer IP"
+    }
+
+    fn extract<T>(&self, req: &http::Request<T>) -> Result<Self::Key, GovernorError> {
+        req.extensions()
+            .get::<TcpConnectInfo>()
+            .and_then(|info| info.remote_addr())
+            .map(|addr| addr.ip())
+            .ok_or(GovernorError::UnableToExtractKey)
+    }
+
+    fn key_name(&self, key: &Self::Key) -> Option<String> {
+        Some(key.to_string())
+    }
+}
+
+/// Convert errors from the Governor rate limiter layer to tonic statuses.
+pub fn tonic_error(e: BoxError) -> tonic::Status {
+    if e.is::<GovernorError>() {
+        // It shouldn't be possible for this to panic, since we already know it's a GovernorError
+        let error = e.downcast_ref::<GovernorError>().unwrap().to_owned();
```

Comment:
You can use downcast_ref in the if clause instead

Reply:
Thanks!